docs: Add Wave api rate limit guide

christopher-hakkaart opened the pull request:
- Add Wave api rate limit guide

munishchouhan commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,155 @@
+
+## How Wave rate limits work
+
+Wave applies rate limits to manifest requests, not layer requests. With an access token, you can pull 2,000 containers per minute. When you pull a container:
```
I think we only need to say, that Wave rate limit applies as per container pull and not each container layer. no need to specify manifest.

christopher-hakkaart replied:
Confirming I've changed this to "Wave applies rate limits to container pulls."

christopher-hakkaart replied:
Changed to "...pulls (manifest requests)..."

pditommaso commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,130 @@
+
+### Subsequent pipeline runs
+
+When you run your pipeline again:
+
+1. The Nextflow head job contacts Wave to check for existing images
+1. Wave finds the cached images (matched by content hash)
+1. Wave returns the registry URLs immediately without rebuilding
+1. All container pulls go directly to your registry
+1. No Wave API calls occur during task execution
+
```
This is partially correct, the pull limit applies also when requesting the freeze container, tho same container will serve many job requests 

https://github.com/seqeralabs/wave/blob/479f50ef9dfd3a630811f6c3dcb0bbe5ad79d280/src/main/groovy/io/seqera/wave/controller/ContainerController.groovy#L266-L267

christopher-hakkaart replied:
Thanks Paolo, folding it into changes now.

christopher-hakkaart commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,130 @@
+**Building without freeze:**
+
+1. Each task requests a manifest from Wave
+1. Wave retrieves the base image from the source registry
```
Good shout.

christopher-hakkaart commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,130 @@
+If an access token isn't provided, the following rate limits apply:
+
+- 25 container builds per day
+- 100 container pulls per hour
```
```suggestion
- 100 container pulls (manifests) per hour
```

christopher-hakkaart commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,130 @@
+If an access token is provided, the following rate limits apply:
+
+- 250 container builds per hour
+- 2,000 container pulls per minute
```
Added 👍

christopher-hakkaart commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,130 @@
+If an access token is provided, the following rate limits apply:
+
+- 250 container builds per hour
+- 2,000 container pulls (manifests) per minute
```
```suggestion
- 2,000 container pulls (manifest requests) per minute
```

christopher-hakkaart commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,134 @@
+
+**Not recommended**: Private Docker Hub for AWS Batch workloads
+
+Private Docker Hub has the following limitations:
```
```suggestion
External container registries have following limitations:
```

christopher-hakkaart commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,134 @@
+- Additional security overhead
+- More complex authentication setup
+
+If you use private Docker Hub, configure Docker credentials on each compute instance.
```
```suggestion
If you use external container registries, configure your credentials on each compute instance.
```

christopher-hakkaart commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,130 @@
+- Lowest latency for AWS workloads
+- Simplest setup
+
+**Not recommended**: Private Docker Hub for AWS Batch workloads
```
I've softened to "**Not recommended**: External container registries for AWS Batch workloads" as per your other suggestion.

Are you suggesting we don't specifically call out AWS? And just refer to "native cloud container registries" instead of "AWS Batch workloads" for both recommended and not recommended?

christopher-hakkaart commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,130 @@
+
+With freeze enabled, Wave is removed from the container pull path. Your compute instances pull directly from your registry with no Wave API calls.
+
+## Configure Wave freeze
```
Later on we talk about the Fusion layer. I have included it in the config, commented it as recommended, and added a note to say why it's recommended but not required.

christopher-hakkaart commented:
Thanks everyone - sorry for all the noise on the PR.

I've made the suggested changes, and a larger restructure after Adams' comments made me realize we had sections that were largely similar (**Building without Wave freeze**, **Building with Wave freeze**, **First pipeline run**, and **Subsequent pipeline runs**).

This has now been shaped into two sections (**Building without Wave freeze** and **Building with Wave freeze**).

There are two open comments.

christopher-hakkaart commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,130 @@
+
+With thousands of concurrent tasks, this approach exceeds rate limits.
+
+**Building with freeze:**
```
I've chatted to @munishchouhan and confirmed this is correct.

I've added a reminder that frozen images will only be reused as long as the image remains the same, and a new section to recommend using specific version tags or SHA256 digests with Wave freeze.

christopher-hakkaart commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,145 @@
+
+## How Wave pull rate limits work
+
+When you pull a container image:
```
```suggestion
When you pull a container image:

1. The Docker service downloads the container manifest (a JSON file detailing each layer of the final image).
1. The Docker service downloads each layer listed in the manifest file.

Wave defines a pull as downloading both the container manifest and all layers. Therefore:
```

gwright99 commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,154 @@
+
+1. Each task requests a manifest from Wave.
+1. Wave performs one of the following actions:
+    - Retrieves the base image manifest from the source registry
```
I'm not sure I fully understand this first bullet. Doesn't building entail having to add extra packages to a base image?  If base image only is being pulled and returned this seems more like the augmentation flow?

gwright99 commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,154 @@
+1. Wave injects the Fusion layer to the container image manifest.
+1. Wave stores the final manifest on Seqera infrastructure.
+1. Wave returns the modified manifest.
+1. Every task creates one API call to Wave.
```
I would lose this bullet -- you have a good description of how we charge API calls above, and this now mixes build mechanics with how we rate-limit.

gwright99 reviewed (changes requested):
In general good. I'm requesting a few prose additions because of events that came up post the session that spawned this original documentation.

christopher-hakkaart commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,154 @@
+- **Azure**: Azure Container Registry (ACR)
+- **Google Cloud**: Google Artifact Registry
+
+**Not recommended**: Third-party container registries.
```
I'll keep the text here minimalist and reframe as "alternate", and add text to highlight the additional effort and public endpoints.

christopher-hakkaart commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,154 @@
+1. Wave finds the frozen images in your registry (matched by content hash).
+1. Wave returns the container URLs in the destination container registry without rebuilding.
+1. All tasks pull the image directly from your registry.
+1. The frozen image serves many task executions.
```
Yup - good call.

christopher-hakkaart commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,154 @@
+When you run your pipeline with Wave freeze for the first time:
+
+1. The Nextflow head job sends your build request to Wave.
+1. Wave checks whether the requested images already exist.
```
Can I have some more context here please? Is there a recording or other thread I can review?

@robsyme @gavinelder @gwright99

gavinelder commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,142 @@
+---
+title: Reducing Wave API calls
```
I would change this title to something like  `Wave Freeze` and have the reduction of API calls be a secondary effect.

The benefit of wave freeze is enhanced security , availability and reliability by reducing the need for wave as part of a pipeline build, acting as a request proxy.

christopher-hakkaart commented:
**Summary of changes**

Changed:

- Removed API limits section from the guide, but added an admonition to link it to the api page.
- Moved the How Wave pull rate limits work to the API page
- Added hierarchy to the API page to sensibly incorporate the sections above

Not changed:

- Making the page "Wave Freeze" focused
    - I would prefer to keep this guide focused on how to avoid API limits rather than making it into a conceptual page about Wave Freeze.
    - Benefits of Wave freeze should be clearly documented, but I want to keep the scope of this guide narrow
    - Wave freeze conceptual content can be addressed as a part of conceptual content in H1 2026.
    
cc @gavinelder

gavinelder commented on docs/nextflow/reduce-api-calls.md:
```diff
@@ -0,0 +1,110 @@
+
+## How Wave freeze avoids rate limits
+
+Wave freeze builds your container image once and stores it in your registry. After the initial build, the source of the container image manifest and layers are redirected to your private registry by Wave.
```
I Quite like the freeze description we have in the wave Q&A Draft google doc as it covers some additional value adds of using wave freeze which may remove the need to draw comparison between wave request flows (Build , Augment , Freeze & Mirror)

christopher-hakkaart commented on docs/api.md:
Awesome!

Resolving to help tidy the PR

christopher-hakkaart commented:
Thanks Gavin - I've pushed more changes. Some were more complex so I've pushed them directly to the PR:

Changed: 

- Removed duplicates of singular authentication/configuration point
- Added more context from the Q&A about Wave Freeze characteristics
	- I'm reluctant to add much more than this here as it should be a conceptual page. This is focused on Wave API rates. 
- Mirrored the structure for first and subsequent builds

Not changed:

- Container registry selection section
    - My view here is that this is a part of the decision process, as there are configuration considerations for third-party container registries and we don't point this out anywhere else in the docs
    
Open questions:

> There is also some additional items where the headjob does this task and at then informs tasks of the local image storage location that workers do not need to communicate with wave directly.

- I'm not sure what to add to address this

gavinelder reviewed (approved):
Overall this looks good to me. 

I have some small editorial notes but they should be considered non blocking and optional.
